### PR TITLE
Fix RpcEvent lifetime

### LINKED ISSUE #788: RpcEvents are not always deleted.
Asan reports cases where an RpcEvent is created but not deleted.
For instance, cartographer_grpc::framework::Rpc::RequestNextMethodInvocation
creates a new RpcEvent as a tag to pass to grpc::ServerContext::AsyncNotifyWhenDone,
but this does not always return the tag to our event queue
(so we cannot delete it).

Here is the stack trace:
```
Direct leak of 160 byte(s) in 5 object(s) allocated from:
    #0 0x60be40 in operator new(unsigned long) ( cartographer/cartographer_grpc.framework.server_test+0x60be40)
    #1 0x62e2c2 in cartographer_grpc::framework::Rpc::RequestNextMethodInvocation()  cartographer/cartographer_grpc/framework/rpc.cc:83:7
    #2 0x6685ee in cartographer_grpc::framework::Service::StartServing(std::vector<cartographer_grpc::framework::CompletionQueueThread, std::allocator<cartographer_grpc::framework::CompletionQueueThread> >&, cartographer_grpc::framework::ExecutionContext*)  cartographer/cartographer_grpc/framework/service.cc:52:7
    #3 0x63ec12 in cartographer_grpc::framework::Server::Start()  cartographer/cartographer_grpc/framework/server.cc:130:5
    #4 0x60f56a in cartographer_grpc::framework::(anonymous namespace)::ServerTest_StartAndStopServerTest_Test::TestBody()  cartographer/cartographer_grpc/framework/server_test.cc:153:3
```

A possible fix is that
- Rpc owns a number of RpcEvents for the different callbacks by grpc
- The RpcEvent with Event::WriteNeeded, which needs to be created dynamically, is not passed to grpc but directly to the event queue and deleted after being processed.

### PR DESCRIPTION
**Pull request by gaschler**
Fixes #788.

Uses two different types of events whether the event goes through the CompletionQueue or not.
CompletionQueueRpcEvent is again a member of Rpc.

### CODE REVIEW

**Comment by gaschler** [comment]
I also tried to make EventQueue hold unique_ptr with two different deleters. However, EventQueue sometimes returns a nullptr, and thereby assigns no deleter, so this approach did not work.

Other alternatives would be to
- make a shallow copy of CompletionQueueRpcEvent when it goes into the EventQueue that may be deleted
- do some renaming and move the "delete this" somewhere less surprising (like, have a method DeleteIfNeeded)

**Comment by cschuet** [comment]
> I also tried to make EventQueue hold unique_ptr with two different deleters. However, EventQueue sometimes returns a nullptr, and thereby assigns no deleter, so this approach did not work.

The trick is to
1. Use a single deleter with a boolean member indicating whether to perform the deletion.
2. Make the deleter default constructable, so that the unique_ptr can be implicitly created from the `return nullptr;` statements in BlockingQueue.

Here is a PR against your repo that does that https://github.com/gaschler/cartographer/pull/1

File: cartographer_grpc/framework/rpc.h [hunk 1]
```diff
@@ -50,11 +48,58 @@ class Rpc {
...
+
     const Event event;
-    std::weak_ptr<Rpc> rpc;
+  };
+
+  class EventDeleter {
+   public:
+    enum Action { DELETE = 0, DO_NOT_DELETE };
+
+    EventDeleter() : action_(DELETE) {}
```
Comment (cschuet) [inline cartographer_grpc/framework/rpc.h]: Add comment why we need this guy?

**Comment by gaschler** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user gaschler. Merge queue now has a length of 1.

**Pull request merged**